internal/data/common.go

```diff
@@ -140,6 +142,136 @@ func NewAttributes(m AttributesMap, droppedCount uint32) Attributes {
+}
+
+func newStringMap(orig *[]*otlpcommon.StringKeyValue) StringMap {
+	return StringMap{orig}
+}
+
+// NewStringMap creates a new StringMap from the given map[string]AttributeValue.
+func NewStringMap(attrMap map[string]string) StringMap {
+	if len(attrMap) == 0 {
+		var orig []*otlpcommon.StringKeyValue
+		orig = nil
+		return StringMap{&orig}
+	}
+	origs := make([]otlpcommon.StringKeyValue, len(attrMap))
+	wrappers := make([]*otlpcommon.StringKeyValue, len(attrMap))
+
+	ix := 0
+	for k, v := range attrMap {
+		wrappers[ix] = &origs[ix]
+		wrappers[ix].Key = k
+		StringKeyValue{wrappers[ix]}.SetValue(v)
+		ix++
+	}
+
+	return StringMap{&wrappers}
+}
+
+// Get returns the StringKeyValue associated with the key and true,
+// otherwise an invalid instance of the StringKeyValue and false.
+func (sm StringMap) Get(k string) (StringKeyValue, bool) {
+	for _, a := range *sm.orig {
+		if a.Key == k {
+			return StringKeyValue{a}, true
+		}
+	}
+	return StringKeyValue{nil}, false
+}
+
+// Delete deletes the entry associated with the key and returns true if the key
+// was present in the map, otherwise returns false.
+func (sm StringMap) Delete(k string) bool {
+	for i, a := range *sm.orig {
+		if a.Key == k {
+			(*sm.orig)[i] = (*sm.orig)[len(*sm.orig)-1]
+			*sm.orig = (*sm.orig)[:len(*sm.orig)-1]
+			return true
+		}
+	}
+	return false
+}
+
+// Insert adds the StringKeyValue to the map when the key does not exist.
+// No action is applied to the map where the key already exists.
+func (sm StringMap) Insert(k, v string) {
+	if _, existing := sm.Get(k); !existing {
+		*sm.orig = append(*sm.orig, NewStringKeyValue(k, v).orig)
+	}
+}
+
+// Update updates an existing StringKeyValue with a value.
+// No action is applied to the map where the key does not exist.
+func (sm StringMap) Update(k, v string) {
+	if av, existing := sm.Get(k); existing {
+		av.SetValue(v)
+	}
+}
+
+// Upsert performs the Insert or Insert action. The StringKeyValue is
+// insert to the map that did not originally have the key. The key/value is
+// updated to the map where the key already existed.
+func (sm StringMap) Upsert(k, v string) {
+	if av, existing := sm.Get(k); existing {
+		av.SetValue(v)
+	} else {
+		*sm.orig = append(*sm.orig, NewStringKeyValue(k, v).orig)
+	}
+}
+
+// Len returns the number of StringKeyValue in the map.
+func (sm StringMap) Len() int {
+	return len(*sm.orig)
+}
+
+// GetStringKeyValue returns the StringKeyValue associated with the given index.
+//
+// This function is used mostly for itereting over all the values in the map:
+// for i := 0; i < am.Len(); i++ {
+//     akv := am.GetStringKeyValue(i)
+//     ... // Do something with the attribute
+// }
+func (sm StringMap) GetStringKeyValue(ix int) StringKeyValue {
+	return StringKeyValue{(*sm.orig)[ix]}
+}
+
+// Sort sorts the entries in the StringMap so two instances can be compared.
+// Returns the same instance to allow nicer code like:
+// assert.EqualValues(t, expected.Sort(), actual.Sort())
```

Comment:
Nice!

Reply:
Is it any convention for functions that are mainly for testing: `SortForTest` or something? I remember something like that but cannot find :(

Reply:
I don't know. Fine to keep the name `Sort`.